### ISSUE TITLE
Missing files

### ISSUE THREAD

**Issue by mirabilos** [zevv/bucklespring#21]
```
Error opening audio file "./wav/7f-1.wav": Failed to open file
Error opening audio file "./wav/7f-0.wav": Failed to open file
Error opening audio file "./wav/56-1.wav": Failed to open file
Error opening audio file "./wav/56-0.wav": Failed to open file
```

**Comment by zevv** [comment]
Which keys on your keyboard are you pressing when these messages are printed?

**Comment by mirabilos** [comment]
Would be easier if they were printed more than once… well, let me try.

    Pause/Break ⇒ Error opening audio file "./wav/77-1.wav": Failed to open file

Should be straight-forward, as it also exists on the Model Clickety-Click.

    The 102nd key ⇒ Error opening audio file "./wav/56-1.wav": Failed to open file
                    Error opening audio file "./wav/56-0.wav": Failed to open file

This key also exists on the international versions of Model M, it’s usually labelled “< > |”. On my keyboard layout, it’s used as Meta key, but as you use scancodes, not keycodes (you d̲o̲ use scancodes, not keycodes, right?), the keyboard mapping used in X11 ought to not make any difference.

    Windows Menu key ⇒ Error opening audio file "./wav/7f-1.wav": Failed to open file
                       Error opening audio file "./wav/7f-0.wav": Failed to open file

Now this is one key which did not exist back then, and does not exist on my own Model M. (Mapped to Compose in my keyboard layout.)

Interestingly enough, the two IBM Thinkpad-specific “Page left” and “Page right” keys *do* have sound… huh…

**Comment by zevv** [comment]
Yeah, this is still a bit of a hairy issue: there's always keys which are not on my model M space saver which are on other keyboards. There is the `-f` option to use a fallback sound for unknown keys, but these will always be played from the center, so the panning is usually off.

Ideally I should at least provide samples and location info for the numeric keypad, and probably have default fallbacks enabled for all unknown keys.

About the scan codes vs key codes: I'm actually using keycodes, but the way I receive them from X is a bit fishy. I use the XRecord extension to pass me events of all pressed keys (not keys which are only hooked), but the data which is passed by X is kind of opaque and needs some bit fiddling to unpack. I never looked back after writing this code, but I might have made some wrong assumptions here. For example, line 71 of scan-linux.c still says

````
        key -= 8; /* X code to scan code? */
````

In other words: I have no clue what I'm doing here :)

**Comment by zevv** [comment]
Addendum: it seems that my wild guess of subtracting 8 was partly right, but is only valid for keycodes < 97. I'm not sure where to find the  authoritative documentation on this, but some hints can be found here: https://github.com/kwhat/libuiohook/blob/master/src/x11/input_helper.c

**Comment by mirabilos** [comment]
The problem is that, in addition, the X scancodes differ between input methods, the classical input (XFree86®, X.org keyboard input) basically uses the PC/BIOS scancodes, whereas libev (e.g. X.org from Debian wheezy onwards) uses something that looks like USB keyboard scancodes…

… and yes, they start at 8, but…

… anyway, my .Xmodmap which maps all non-fancy (e.g. no sound volume) keys is [here for classic X](http://www.mirbsd.org/cvs.cgi/contrib/samples/dot.Xmodmap?rev=1.26;content-type=text%2Fplain) and [here for evdev](http://www.mirbsd.org/cvs.cgi/contrib/samples/dot.Xmodmap?rev=1.24.4.8;content-type=text%2Fplain), in case that helps (the LHS especially; do mind that some keys are mapped differently, like Esc and Alt).

**Comment by zevv** [comment]
* On 2016-09-16 20:03:01 +0200, mirabilos wrote:
 
> The problem is that, in addition, the X scancodes differ between input
> methods, the classical input (XFree86®, X.org keyboard input) basically
> uses the PC/BIOS scancodes, whereas libev (e.g. X.org from Debian wheezy
> onwards) uses something that looks like USB keyboard scancodes…
>
> … and yes, they start at 8, but…
>
> … anyway, my .Xmodmap which maps all non-fancy (e.g. no sound volume) keys
> is [1]here for classic X and [2]here for evdev, in case that helps (the
> LHS especially; do mind that some keys are mapped differently, like Esc
> and Alt).

Well, I'm not sure what is the right way to go here. Kind of by accident
buckle has been ported to BSD, MacOS and even Win32, which is probably a
good recipe for opening the gates to hell, scancode-portability-wise. 

I guess the safest and most flexible method will be to use lookup tables
for mapping keys for each platform, or even better, putting this in a
config file so users can easily tweak their setups.

I'm very linux/PC centric myself, but I'm always open to good ideas -
what would be the most portable way to handle this?

-- 
:wq
^X^Cy^K^X^C^C^C^C

**Comment by mirabilos** [comment]
Ico Doornekamp dixit:

>I guess the safest and most flexible method will be to use lookup
>tables for mapping keys for each platform

Not really… these depend on run-time configuration, not (just) on the
platform.

>or even better, putting this in a config file so users can easily
>tweak their setups.

Hmm, “can” doesn’t equal “will” or “will be happy to”…

… but that’s probably the first realistic step.

>I'm very linux/PC centric myself, but I'm always open to good ideas -
>what would be the most portable way to handle this?

Can’t say offhand either… this needs more thought. For now it’s
good enough as it is, though.

>-- 
>:wq
>^X^Cy^K^X^C^C^C^C

Like https://www.gnu.org/fun/jokes/ed-msg.html ? ☻☺

bye,
//mirabilos
-- 
> emacs als auch vi zum Kotzen finde (joe rules) und pine für den einzig
> bedienbaren textmode-mailclient halte (und ich hab sie alle ausprobiert). ;)
Hallooooo, ich bin der Holger ("Hallo Holger!"), und ich bin ebenfalls
... pine-User, und das auch noch gewohnheitsmäßig ("Oooooooohhh").  [aus dasr]

**Comment by AbuMaia** [comment]
I have a similar issue. These errors come up when starting bucklespring, before hitting any keys:

> abumaia@AbuMaia:~$ bucklespring.buckle
Error opening audio file "/snap/bucklespring/23/wav/f0-1.wav": Failed to open file
Error opening audio file "/snap/bucklespring/23/wav/f0-0.wav": Failed to open file

I've looked through the folder. I'm guessing it's a typo somewhere in the code, intended to reference 0f-1 and 0f-0. What do I need to look at to fix it?
